### PR TITLE
8282944: GHA: Add Alpine Linux x86_64 pre-integration check

### PR DESCRIPTION
**Pull request by gdams**
Adds Alpine build CI job to the GitHub actions matrix. Currently this only builds hotspot. I can add tests too if people think that would be useful but for now I've left it as just build.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8282944](https://bugs.openjdk.org/browse/JDK-8282944): GHA: Add Alpine Linux x86_64 pre-integration check (**Enhancement** - P4)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20577/head:pull/20577` \
`$ git checkout pull/20577`

Update a local copy of the PR: \
`$ git checkout pull/20577` \
`$ git pull https://git.openjdk.org/jdk.git pull/20577/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20577`

View PR using the GUI difftool: \
`$ git pr show -t 20577`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20577.diff">https://git.openjdk.org/jdk/pull/20577.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20577#issuecomment-2288223472)

### CODE REVIEW

**Comment by magicus** [comment]
The GHA code itself looks basically good. I was thinking a bit about the name, I wonder if "linux-alpine-x64" would not be better, to keep the os name in front (after all, Alpine is just a flavor of Linux, not a completely new OS). But maybe that's just me..?

The big question, as always with GHA, is if there is enough interest from the community to warrant adding this configuration to GHA. GHA testing is done on the developer's individual account's quota. It takes time and resources, and since we've introduced it we've basically only added platforms, never removing them.

Can you elaborate a bit what problems you expect to find with this addition, that is not currently detected by GHA, and why this is important enough to always run this in GHA? Basically, convince me that it is worth all the computing power and everyone's time to add this (what to me looks like a) special case.

**Comment by gdams** [comment]
> The big question, as always with GHA, is if there is enough interest from the community to warrant adding this configuration to GHA. GHA testing is done on the developer's individual account's quota. It takes time and resources, and since we've introduced it we've basically only added platforms, never removing them.
> 
> Can you elaborate a bit what problems you expect to find with this addition, that is not currently detected by GHA, and why this is important enough to always run this in GHA? Basically, convince me that it is worth all the computing power and everyone's time to add this (what to me looks like a) special case.

Including Alpine Linux in the GHA pipeline is valuable because it uses the musl C library, which is different from the glibc found in most other Linux distributions. This difference can expose compatibility issues that might not be caught elsewhere. Given Alpine's popularity in containerised environments due to its small footprint and security features, ensuring that OpenJDK works well on Alpine helps meet the needs of developers who rely on these setups. By testing on Alpine, we can catch potential issues early and make OpenJDK more reliable in a variety of environments, especially in cloud-native applications where Alpine is often used.

**Comment by magicus** [comment]
Thanks. From my point of view, the code looks okay. However, I'd like to hear someone else weigh in on the question if this should be enabled or not.

**Comment by shipilev** [comment]
Honestly, I am ambivalent about this. This only captures the Hotspot build errors, which I think are rare for Alpine, judging by looking around the JBS. And it makes sense: the build failures looks basically around compiling against weird libc APIs that might not be supported by musl. I think most of the issues with Alpine are runtime problems in tests?

So it looks that building only Hotspot is in weird spot: we spend time without getting much of the testing off it. How bad would it be to run testing with Alpine? That would probably not be very well received, though, if no one could debug Alpine test failures...

Tangentially, I think we should start thinking about extended workflows that are _not_ run by default PR-triggers, so you could potentially run them if you wanted for Alpine-specific fixes. This would making the decisions like this rather easy.

**Comment by gdams** [comment]
> Honestly, I am ambivalent about this. This only captures the Hotspot build errors, which I think are rare for Alpine, judging by looking around the JBS. And it makes sense: the build failures looks basically around compiling against weird libc APIs that might not be supported by musl. I think most of the issues with Alpine are runtime problems in tests?
> 
> So it looks that building only Hotspot is in weird spot: we spend time without getting much of the testing off it. How bad would it be to run testing with Alpine? That would probably not be very well received, though, if no one could debug Alpine test failures...
> 
> Tangentially, I think we should start thinking about extended workflows that are _not_ run by default PR-triggers, so you could potentially run them if you wanted for Alpine-specific fixes. This would making the decisions like this rather easy.

I was actually going to try building the whole JDK and see what the difference in build time was. If it's not significant then it might be a good first step to just add full build testing. We could always follow up with test running at a later stage if people feel it would be important?

**Comment by gdams** [comment]
For reference:

| Build Type | Duration |
| ----------  | -------- |
| hotspot (debug) | 21m |
| hotspot (release) | 15m |
| full (debug) | TBD |
| full (release) | TBD |

**Comment by magicus** [comment]
> Tangentially, I think we should start thinking about extended workflows that are not run by default PR-triggers, so you could potentially run them if you wanted for Alpine-specific fixes. This would making the decisions like this rather easy.

I agree, we should really add that functionality. Then, as you say, it would be a no-brainer to add Alpine as an opt-in build and/or build+test.

If you are hesitant as well about adding this, maybe this is something we should prioritize before accepting this. (And then only accepting this as a non-default option.)

It might be so simple as to not list it in the `default` field of the platform selection part. 

@gdams Maybe you can test that? Revert line 39 in main.yml, and start the test manually and specify alpine as a target.

If that works fine, we can add some way for users to set their own defaults by setting a GHA environment/repository variable. (Either an "JDK_EXTRA_TARGETS" to just appent more to the default, or a "JDK_TARGETS" to completely override, or both.)

**Comment by gdams** [comment]
> > Tangentially, I think we should start thinking about extended workflows that are not run by default PR-triggers, so you could potentially run them if you wanted for Alpine-specific fixes. This would making the decisions like this rather easy.
> 
> I agree, we should really add that functionality. Then, as you say, it would be a no-brainer to add Alpine as an opt-in build and/or build+test.
> 
> If you are hesitant as well about adding this, maybe this is something we should prioritize before accepting this. (And then only accepting this as a non-default option.)
> 
> It might be so simple as to not list it in the `default` field of the platform selection part.
> 
> @gdams Maybe you can test that? Revert line 39 in main.yml, and start the test manually and specify alpine as a target.
> 
> If that works fine, we can add some way for users to set their own defaults by setting a GHA environment/repository variable. (Either an "JDK_EXTRA_TARGETS" to just appent more to the default, or a "JDK_TARGETS" to completely override, or both.)

I'm not actually sure if line 39 is where we want to make the changes. I think we need to add some logic to this function to skip certain platforms by default when a push event is triggered:

https://github.com/openjdk/jdk/blob/c6906eff9db8f7e7737c42af49ae4f2f6ad7b095/.github/workflows/main.yml#L77-L112

**Review by magicus — APPROVED** [review]
For the future, I think we should try to figure out a way to allow users to define their own set of included checks persistently, but for now this is good. Since it will not run by default, I have no objections.

**Comment by gdams** [comment]
> For the future, I think we should try to figure out a way to allow users to define their own set of included checks persistently, but for now this is good. Since it will not run by default, I have no objections.

I think this is the idea of these `JDK_SUBMIT_FILTER` and `JDK_SUBMIT_PLATFORMS`:

https://github.com/openjdk/jdk/blob/4a6cbe47daae82970c36c8cd86655581b2c59f59/.github/workflows/main.yml#L86-L92

Admittedly with my current setup this wouldn't work for Alpine. The only other option I could think of is maintaining a list of "skip unless explicitly defined" platforms in the `check_platform()` function. I'd be happy to put a prototype together if you'd prefer this solution

**Comment by gdams** [comment]
@magicus I think I've got this working correctly now. If users choose to they can run alpine testing by default by following https://wiki.openjdk.org/display/skara/testing#Testing-Conditionallyenable/disablebuildsandtests or run it manually by triggering the `workflow_dispatch` but by default Alpine will not run

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -78,6 +79,10 @@ jobs:
           # Returns 'true' if the input platform list matches any of the platform monikers given as argument,
           # 'false' otherwise.
           # arg $1: platform name or names to look for
+
+          # List of platforms to exclude by default
+          EXCLUDED_PLATFORMS=("alpine-linux-x64")
```
Comment (magicus) [inline .github/workflows/main.yml:84]: I'm not to fond of declaring a list like this in the middle of a code body. Perhaps you can set it up as a non-required workflow input with alpine as the default value, declared next to the default list of platforms?

**Comment by magicus** [comment]
Yes, this is somewhat more akin to what I was thinking. But I was actually thinking about a more encompassing rewrite of that method.

First of all, GHA variables did not exist when that code were written, so they had to be secrets, even though it does not make sense in this context. These control variables should be changed from secrets to variables, or the code rewritten so it can support both.

Secondly, I think there is a use-case for having both a variable which just appends to the default set of platforms/tests, and one that replaces the default value.

Third, there has been requests to change what tests are run  (e.g. tier-2) as well as which platforms are tested. I think it would be good to let such functionality tie in with any other changes in this area.

But, just to be clear, I don't think it is necessary, nor even appropriate, for you to address any of these changes in this PR. In fact, I would have been happy with the previous iteration as well. But this version is okay as well, if you can move the exclude list to a more suited location.

File: .github/workflows/main.yml (hunk 1 above)
Comment (gdams) [reply .github/workflows/main.yml:84]: yeah this makes sense, I'll make that change now

**Comment by gdams** [comment]
> Yes, this is somewhat more akin to what I was thinking. But I was actually thinking about a more encompassing rewrite of that method.
> 
> First of all, GHA variables did not exist when that code were written, so they had to be secrets, even though it does not make sense in this context. These control variables should be changed from secrets to variables, or the code rewritten so it can support both.
> 
> Secondly, I think there is a use-case for having both a variable which just appends to the default set of platforms/tests, and one that replaces the default value.
> 
> Third, there has been requests to change what tests are run (e.g. tier-2) as well as which platforms are tested. I think it would be good to let such functionality tie in with any other changes in this area.
> 
> But, just to be clear, I don't think it is necessary, nor even appropriate, for you to address any of these changes in this PR. In fact, I would have been happy with the previous iteration as well. But this version is okay as well, if you can move the exclude list to a more suited location.

yeah I totally agree with this idea, I'm happy to make these changes but that's outside of the scope of this PR, I'll test some changes in my local environment.

File: .github/workflows/main.yml (hunk 1 above)
Comment (gdams) [reply .github/workflows/main.yml:84]: @magicus do you want me to push this change?

**Pull request closed without merging**